ampd-handlers/src/bin/solana/handler.rs

```diff
@@ -0,0 +1,981 @@
+            futures::future::join_all(tx_calls)
+                .await
+                .into_iter()
+                .flatten()
+                .collect::<HashMap<Signature, SolanaTransaction>>();
+
+        Ok(finalized_tx_receipts)
+    }
+}
+
+#[async_trait]
+impl<C> EventHandler for Handler<C>
+where
+    C: SolanaRpcClientProxy + Send + Sync,
+{
+    type Err = Error;
+    type Event = PollStartedEvent;
+
+    async fn handle<HC: EventHandlerClient + Send + 'static>(
+        &self,
+        event: PollStartedEvent,
+        client: &mut HC,
+    ) -> Result<Vec<cosmrs::Any>> {
+        let (source_gateway_address, poll_id) = match event {
+            PollStartedEvent::Messages(ref message) => {
+                (&message.source_gateway_address, &message.poll_id)
+            }
+            PollStartedEvent::VerifierSet(ref verifier_set) => {
+                (&verifier_set.source_gateway_address, &verifier_set.poll_id)
+            }
+        };
+
+        // Validate that the source gateway address matches the configured gateway address
+        if *source_gateway_address != self.gateway_address {
+            info!(
+                poll_id = poll_id.to_string(),
+                expected_gateway = %self.gateway_address,
+                actual_gateway = %source_gateway_address,
+                "skipping poll due to gateway address mismatch"
+            );
+            return Ok(vec![]);
+        }
+
+        VotingHandler::handle(self, event.into(), client).await
+    }
+
+    fn subscription_params(&self) -> SubscriptionParams {
+        let attributes = HashMap::from([(
+            "source_chain".to_string(),
+            serde_json::Value::String(self.chain.to_string()),
+        )]);
+
+        SubscriptionParams::new(
+            vec![
+                AbciEventTypeFilter {
+                    event_type: MessagesPollStarted::event_type(),
+                    contract: self.voting_verifier_contract.clone(),
+                    attributes: attributes.clone(),
+                },
+                AbciEventTypeFilter {
+                    event_type: VerifierSetPollStarted::event_type(),
+                    contract: self.voting_verifier_contract.clone(),
+                    attributes,
+                },
+            ],
+            false,
+        )
+    }
+}
+
+impl<C> Debug for Handler<C>
+where
+    C: SolanaRpcClientProxy + Send + Sync,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Handler")
+            .field("verifier", &self.verifier)
+            .field("voting_verifier_contract", &self.voting_verifier_contract)
+            .field("gateway_address", &self.gateway_address)
+            .field("domain_separator", &self.domain_separator)
+            .field("rpc_client", &"WARN: Solana sdk does impl Debug")
+            .field("monitoring_client", &self.monitoring_client)
+            .finish()
+    }
+}
```

Comment:
I think I would prefer to just except the handler from the debug log in `handle_event` in event_handler.rs by changing the instrumentation line to `#[instrument(skip(self), fields(handler_task_config = ?self.config))]`. Then it won't require the handler to implement the debug trait and we don't have to do this manual implementation